File: php/quick_sort.php
<?php

function quick_sort($array, $size = null)
{
    if ($size === null) {
        $size = count($array);
    }

    if ($size <= 1) {
        return $array;
    }

    // select pivot (median of three)
    $first = $array[0];
    $last = $array[$size - 1];
    if ($first > $last) {
        $temp = $first;
        $first = $last;
        $last = $temp;
    }

    if ($size === 2) {
        return [$first, $last];
    }

    $pivot = $array[ (int)($size/2) ];
    if ($pivot < $first) {
        $pivot = $first;
    } elseif ($pivot > $last) {
        $pivot = $last;
    }
    
    // moves values
    $left = [];
    $middles = [];
    $right = [];
    foreach ($array as $value) { // a for loop is 50% slower 
        if ($value < $pivot) {
            $left[] = $value;
        } elseif ($value > $pivot) {
            $right[] = $value;
        } else {
            $middles[] = $value;
        }        
    }

    $leftSize = count($left);
    // creating leftSize above and incrementing if in the loop 
    // is noticeably slower than using count()
    if ($leftSize > 1) {
        $left = quick_sort($left, $leftSize);
    }

    $rightSize = count($right);
    if ($rightSize > 1) {
        $right = quick_sort($right, $rightSize);
    }

    return array_merge($left, $middles, $right);
}

// ----------------------------------------
// tests / stats

// check that the array is indeed sorted
function check($array)
{
    $count = count($array);
    for ($i=1; $i < $count; $i++) { 
        if ($array[$i-1] > $array[$i]) {
            return false;
        }
    }
    return true;
}

$arrayCount = (int)$argv[1];
$arraySize = (int)$argv[2];

$data = [];
srand(time());
rand(); rand(); rand();

// build an array of $arrayCount arrays of $arraySize size
for ($i = 0; $i < $arrayCount; $i++) { 
    $a = [];
    for ($j = 0; $j < $arraySize; $j++) { 
        $a[] = rand(0, $arraySize);
    }
    $data[] = $a;
}

$starTime = microtime(true);

for ($i = 0; $i < $arrayCount; $i++) { 
    // if (!check(quick_sort($data[$i]))) {
    //     echo "wrong";
    // }
    quick_sort($data[$i]);
}

$endPhpMergeTime = microtime(true);
$phpMergeDiff = $endPhpMergeTime - $starTime;

for ($i = 0; $i < $arrayCount; $i++) { 
    // if (!check(Sort\QuickSort::make($data[$i]))) {
    //     echo "wrong2";
    // }
    Sort\QuickSort::make($data[$i]);
}

$endExtMergeTime = microtime(true);
$extMergeDiff = $endExtMergeTime - $endPhpMergeTime;

for ($i = 0; $i < $arrayCount; $i++) { 
    sort($data[$i]);
}

$endSortTime = microtime(true);
$phpSortDiff = $endSortTime - $endExtMergeTime;

// execute the C script and retrive the last line of its input (the one that begins by "C")
$cTime = exec("./c/builds/quick_sort $arrayCount $arraySize | grep C");
// calculating C time with PHP whould gives a time slightly higher

$str = <<<EOL
<pre>
Array count: $arrayCount
Array size: $arraySize
php userland:           $phpMergeDiff s
php extension (zephir): $extMergeDiff s
php built-in sort:      $phpSortDiff s
$cTime
</pre>
EOL;
echo "$str";

/*
Array count: 50
Array size: 10000
php userland:           0.80580806732178 s
php extension (zephir): 0.35681986808777 s
php built-in sort:      0.064460039138794 s
C:                      0.072400 s

Array count: 30
Array size: 100000
php userland:           5.7532091140747 s
php extension (zephir): 2.5272769927979 s
php built-in sort:      0.4820020198822 s
C:                      0.526823 s

Array count: 10
Array size: 500000
php userland:           10.721113920212 s
php extension (zephir): 4.752336025238 s
php built-in sort:      0.93213200569153 s
C:                      0.981925 s

Array count: 5
Array size: 1000000
php userland:           11.271620988846 s
php extension (zephir): 5.0043919086456 s
php built-in sort:      0.98648905754089 s
C:                      1.029710 s
*/

File: c/quick_sort.c
#include <stdlib.h>
#include <stdio.h>
#include <time.h>


void display(int array[], int size, char *text)
{
    int i;
    printf("%s size=%d : ", text, size);
    for (i = 0; i < size; i++) {

        printf("%d ", array[i]);
    }
    printf("\n");
}


void quick_sort(int *array, const uint size)
{
    if (size <= 1) {
        return;
    }

    // select pivot (median of three)
    int first = array[0];
    int last = array[size - 1];
    if (first > last) {
        int temp = first;
        first = last;
        last = temp;
    }

    if (size == 2) {
        array[0] = first;
        array[1] = last;
        return;
    }

    int pivot = array[ (int)(size/2) ];
    if (pivot < first) {
        pivot = first;
    } else if (pivot > last) {
        pivot = last;
    }

    // moves values
    // could improve memory consumption by counting the values first
    // so that to create left and right with just the correct size
    // but it's 50% slower...
    // maybe we could realloc by chunk of 1/4 size ony when we need it...
    int *left  = malloc(sizeof(int) * size);
    int *right = malloc(sizeof(int) * size);

    int i, leftSize = 0, middleSize = 0, rightSize = 0;
    for (i = 0; i < size; i++) {
        int value = array[i];
        if (value < pivot) {
            left[ leftSize++ ] = value;
        } else if (value > pivot) {
            right[ rightSize++ ] = value;
        } else {
            middleSize++;
        }        
    }

    if (leftSize > 1) {
        quick_sort(left, leftSize);
    }

    if (rightSize > 1) {
        quick_sort(right, rightSize);
    }

    // merge
    for (i = 0; i < leftSize; i++) {
        array[i] = left[i];
    }
    for (i = 0; i < middleSize; i++) {
        array[ (leftSize + i) ] = pivot;
    }
    const int offset = leftSize + middleSize;
    for (i = 0; i < rightSize; i++) {
        array[ (offset + i) ] = right[i];
    }

    free(left);
    free(right);
}


int check(int *array, int size)
{
    for (int i=1; i < size; i++) { 
        if (array[i-1] > array[i]) {
            return 0;
        }
    }
    return 1;
}


int main(int argc, char *argv[]) 
{
    ulong arrayCount = atoi(argv[1]);
    ulong arraySize = atoi(argv[2]);
    ulong total = arraySize * arrayCount;
    ulong i;
    
    int *data = malloc(sizeof(int) * total);
    srand(time(NULL));
    rand(); rand(); rand();

    for (i = 0; i < total; i++) {
        data[i] = rand() % arraySize;
    }

    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC_RAW, &start);

    for (i = 0; i < arrayCount; i++) 
    {
        int *_data = (data + (i * arraySize));
        // display(_data, arraySize, "BEFORE");
        quick_sort( _data, arraySize);
        // if (! check(_data, arraySize)) {
        //     display(_data, arraySize, "ERROR: ");
        // }
    }

    clock_gettime(CLOCK_MONOTONIC_RAW, &end);
    double diffSec = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1E9;

    printf("Array count: %d \n", (int)arrayCount);
    printf("Array size: %d \n", (int)arraySize);
    printf("C:                      %f s\n", diffSec);

    return 0;
}
